### PR TITLE
[Coop] Convert ves_icall_System_Globalization_CultureData_fill_number…

### PR DESCRIPTION
**Pull request by jaykrell**
…_data.

Tested by running this program:
```
class A
{
	public static void Main()
	{
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.CompareInfo);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.DateTimeFormat.AMDesignator);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.DateTimeFormat.DateSeparator);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.DisplayName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.EnglishName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.IsNeutralCulture);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.IsReadOnly);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.LCID);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.Name);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NativeName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NumberFormat.CurrencySymbol);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalDigits);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalSeparator);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.Parent);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.TextInfo);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.ThreeLetterISOLanguageName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.ThreeLetterWindowsLanguageName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.TwoLetterISOLanguageName);
	}
}
```

which covers some but not all of the data.

### CODE REVIEW

**Comment by jaykrell** [comment]
Moved more to managed per https://github.com/mono/mono/pull/12023#issuecomment-450952025.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by lambdageek** [comment]
@monojenkins build

File: mono/metadata/culture-info.h [hunk 1]
```diff
@@ -53,34 +53,65 @@ typedef struct {
...
+	gint8 number_decimal_digits;
+
+	gint currency_group_sizes [2];
+	gint number_group_sizes [2];
 } NumberFormatEntry;
 
+// Due to the questionable layout of NumberFormatEntry, in particular
+// 7x byte, make something more guaranteed to match between native and managed.
+struct NumberFormatEntryManaged {
+	int currency_decimal_separator;
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/culture-info.h]: `gint32` ok

File: mono/metadata/culture-info.h [hunk 2]
```diff
@@ -53,34 +53,65 @@ typedef struct {
...
+	gint8 currency_decimal_digits;
+	gint8 number_decimal_digits;
+
+	gint currency_group_sizes [2];
+	gint number_group_sizes [2];
 } NumberFormatEntry;
 
+// Due to the questionable layout of NumberFormatEntry, in particular
+// 7x byte, make something more guaranteed to match between native and managed.
+struct NumberFormatEntryManaged {
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/culture-info.h]: ok

File: mcs/class/corlib/ReferenceSources/CultureData.cs [hunk 3]
```diff
@@ -611,7 +611,48 @@ internal static bool IsCustomCultureId(int cultureId)
 			return false;
 		}
 
-		internal void GetNFIValues (NumberFormatInfo nfi)
+		[StructLayout (LayoutKind.Sequential)]
+		internal struct NumberFormatEntry
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mcs/class/corlib/ReferenceSources/CultureData.cs]: agreed

File: mono/metadata/culture-info.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/culture-info.h]: `gint32` is pointless though. `int` is always going to be 32bits on all interesting systems past and present.

File: mono/metadata/culture-info.h [hunk 4]
```diff
@@ -53,34 +53,65 @@ typedef struct {
...
-	const gint8 percent_negative_pattern;
-	const gint8 percent_positive_pattern;
-	const gint8 number_negative_pattern;
-
-	const gint8 currency_decimal_digits;
-	const gint8 number_decimal_digits;
-
-	const gint currency_group_sizes [GROUP_SIZE];
-	const gint number_group_sizes [GROUP_SIZE];	
+	// 12x ushort -- 6 ints
```
Comment (luhenry) [inline mono/metadata/culture-info.h]: This was already discussed. It's exhausting you are trying to submit it in another PR. Revert that  `const` change.

File: mono/metadata/locales.c [hunk 5]
```diff
@@ -284,59 +284,39 @@ ves_icall_System_Globalization_CultureData_fill_culture_data (MonoCultureData *t
...
-	SET_STR (number, perMilleSymbol, domain, idx2string (nfe->per_mille_symbol), error);
-	SET_STR (number, positiveInfinitySymbol, domain, idx2string (nfe->positive_infinity_symbol), error);
-	SET_STR (number, positiveSign, domain, idx2string (nfe->positive_sign), error);
-#undef SET_STR
+	g_assert (index >= 0);
+	NumberFormatEntry const * const native = &number_format_entries [index];
+// We could return the pointer directly, but I'm leary of the 7x byte layout.
+// If the data is regenerated, suggest a byte of padding there.
+#define X(x) managed->x = native->x;
+	X (currency_decimal_separator)
```
Comment (luhenry) [inline mono/metadata/locales.c]: Remove that awful hack, and just be explicit.

File: mono/metadata/culture-info.h (hunk 1 above)
Comment (luhenry) [inline mono/metadata/culture-info.h]: `gint32`.

File: mono/metadata/culture-info.h (hunk 4 above)
Comment (jaykrell) [reply mono/metadata/culture-info.h]: const is very unusual and useless and potentially harmful here. I have almost never seen this pattern, and I have seen recent compiler bugs associated with it.

File: mono/metadata/culture-info.h (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/culture-info.h]: Agreed. There are also relevant test failures in CI. It is nice to see that coverage, will investigate later.

File: mono/metadata/locales.c (hunk 5 above)
Comment (jaykrell) [reply mono/metadata/locales.c]: macros are very useful to automate error-prone tedious repitition, and to save people from double checking such code, but ok.

**Comment by jaykrell** [comment]
A failing test:
The p for percent formating.

```
using System;
using System.Globalization;
using System.Threading;

class A
{
	private const Byte MyByte1 = 42;
	private const Byte MyByte2 = 0;
	private const Byte MyByte3 = 255;
	private const string MyString1 = "42";
	private const string MyString2 = "0";
	private const string MyString3 = "255";
	private static string[] Formats1 = {
		"p", "c",  "d",  "e",  "f",  "g",  "n",  "p", "x" };
	private static string[] Formats2 = {
		"p5", "c5", "d5", "e5", "f5", "g5", "n5", "p5", "x5" };
	private static string[] Results1 = {	"",
					"0", "0.000000e+000", "0.00",
					"0", "0.00", "0.00 %", "0"};
	private static string[] Results1_Nfi = {NumberFormatInfo.InvariantInfo.CurrencySymbol+"0.00",
					"0", "0.000000e+000", "0.00",
					"0", "0.00", "0.00 %", "0"};
	private static string[] Results2 = {	"",
					"00255", "2.55000e+002", "255.00000",
					"255", "255.00000", "25,500.00000 %", "000ff"};
	private static string[] Results2_Nfi = {NumberFormatInfo.InvariantInfo.CurrencySymbol+"255.00000", 
					"00255", "2.55000e+002", "255.00000",
					"255", "255.00000", "25,500.00000 %", "000ff"};

	public static void Main()
	{
		for (int i=0; i < Formats1.Length; i++) {
			Console.WriteLine($"i:{i}");
			Console.WriteLine($"Formats1[i]:{Formats1[i]}");
			Console.WriteLine(MyByte3.ToString(Formats1[i]));
			Console.WriteLine($"Formats2[i]:{Formats2[i]}");
			Console.WriteLine(MyByte3.ToString(Formats2[i]));
		}

		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.CompareInfo);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.DateTimeFormat.AMDesignator);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.DateTimeFormat.DateSeparator);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.DisplayName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.EnglishName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.IsNeutralCulture);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.IsReadOnly);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.LCID);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.Name);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NativeName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NumberFormat.CurrencySymbol);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalDigits);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.NumberFormat.NumberDecimalSeparator);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.Parent);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.TextInfo);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.ThreeLetterISOLanguageName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.ThreeLetterWindowsLanguageName);
		System.Console.WriteLine(System.Globalization.CultureInfo.CurrentUICulture.TwoLetterISOLanguageName);
	}
}
```

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed
